src/hash_generator/dlNormalizer.js

```diff
@@ -0,0 +1,398 @@
+  }
+
+  if (uri.match('^urn:epc:idpat:sgcn:(([0-9]{6}.[0-9]{6})|([0-9]{7}.[0-9]{5})|([0-9]{8}.[0-9]{4})|([0-9]{9}.[0-9]{3})|([0-9]{10}.[0-9]{2})|([0-9]{11}.[0-9]{1})|([0-9]{12}..)).\*$')) {
+    const gs1CompanyPrefix = uri.substring(19, partition);
+    const couponRef = uri.substring(
+      partition + 1, (partition + 1 + (12 - gs1CompanyPrefix.length)),
+    );
+    const rawSGCN = gs1CompanyPrefix + couponRef;
+    return (`https://id.gs1.org/255/${addCheckDigitAndZeroPad(rawSGCN, throwError, 13)}`);
+  }
+
+  if (uri.match('^urn:epc:idpat:cpi:(([0-9]{6}.(%2[3dfDF]|%3[0-9]|%4[1-9A-Fa-f]|%5[0-9Aa]|[0-9A-Z-]){1,24})|([0-9]{7}.(%2[3dfDF]|%3[0-9]|%4[1-9A-Fa-f]|%5[0-9Aa]|[0-9A-Z-]){1,23})|([0-9]{8}.(%2[3dfDF]|%3[0-9]|%4[1-9A-Fa-f]|%5[0-9Aa]|[0-9A-Z-]){1,22})|([0-9]{9}.(%2[3dfDF]|%3[0-9]|%4[1-9A-Fa-f]|%5[0-9Aa]|[0-9A-Z-]){1,21})|([0-9]{10}.(%2[3dfDF]|%3[0-9]|%4[1-9A-Fa-f]|%5[0-9Aa]|[0-9A-Z-]){1,20})|([0-9]{11}.(%2[3dfDF]|%3[0-9]|%4[1-9A-Fa-f]|%5[0-9Aa]|[0-9A-Z-]){1,19})|([0-9]{12}.(%2[3dfDF]|%3[0-9]|%4[1-9A-Fa-f]|%5[0-9Aa]|[0-9A-Z-]){1,18})).\*$')) {
+    const gs1CompanyPrefix = uri.substring(18, partition);
+    const separator = uri.lastIndexOf('.');
+    const cpref = uri.substring(partition + 1, separator);
+    const rawCPI = gs1CompanyPrefix + cpref;
+    return (`https://id.gs1.org/8010/${webUriPercentEncoder(rawCPI)}`);
+  }
+
+  if (uri.match('^urn:epc:idpat:itip:(([0-9]{6}.[0-9]{7})|([0-9]{7}.[0-9]{6})|([0-9]{8}.[0-9]{5})|([0-9]{9}.[0-9]{4})|([0-9]{10}.[0-9]{3})|([0-9]{11}.[0-9]{2})|([0-9]{12}.[0-9]{1})).[0-9]{2}.[0-9]{2}.\*$')) {
+    const gs1CompanyPrefix = uri.substring(19, partition);
+    const itemRef = uri.substring(partition + 1, (partition + 1 + (13 - gs1CompanyPrefix.length)));
+    const rawGTIN = itemRef.substring(0, 1) + gs1CompanyPrefix + itemRef.substring(1);
+    const piece = uri.substring(34, 36);
+    const total = uri.substring(37, 39);
+    return (`https://id.gs1.org/8006/${addCheckDigitAndZeroPad(rawGTIN, throwError)}${piece}${total}`);
+  }
+
+  if (uri.match('^urn:epc:idpat:upui:(([0-9]{6}.[0-9]{7})|([0-9]{7}.[0-9]{6})|([0-9]{8}.[0-9]{5})|([0-9]{9}.[0-9]{4})|([0-9]{10}.[0-9]{3})|([0-9]{11}.[0-9]{2})|([0-9]{12}.[0-9]{1})).\*$')) {
+    const gs1CompanyPrefix = uri.substring(19, partition);
+    const itemRef = uri.substring(partition + 1, (partition + 1 + (13 - gs1CompanyPrefix.length)));
+    const rawGTIN = itemRef.substring(0, 1) + gs1CompanyPrefix + itemRef.substring(1);
+    return (`https://id.gs1.org/01/${addCheckDigitAndZeroPad(rawGTIN, throwError)}`);
+  }
+
+  if (!uri.match('^https?:(\/\/((([^\/?#]*)@)?([^\/?#:]*)(:([^\/?#]*))?))?((([^?#]*)(\/(01|gtin|8006|itip|8010|cpid|414|gln|417|party|8017|gsrnp|8018|gsrn|255|gcn|00|sscc|253|gdti|401|ginc|402|gsin|8003|grai|8004|giai)\/)([0-9]{4}[^\/]+)?[\/]?(([^?\n]*))?(#([^]*))?)|(\/[A-Za-z_-]{10}$))')) {
+    return uri;
+  }
+
+  if (uri.includes('?')) {
+    uri = uri.substring(0, uri.indexOf('?'));
+  }
+
+  uri = uri.split('/gtin/').join('/01/')
```

Comment:
It may be beneficial to factor out a reusable `replace` function for use here instead of long chains. Possibly using a clear map for all expected replacements.